## Finding the ideal Neighborhood in Toronto for starting an Indian Restaurant
-----

In [1]:
import pandas as pd
import requests
import numpy as np
import requests
import folium

In [14]:
pip install lxml

Note: you may need to restart the kernel to use updated packages.


#### Reading the data to a dataframe from wikipedia page

In [2]:
df = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')[0]
df.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

#### Deleting rows having Borough as "Not assigned"

In [3]:
df.drop(df[df['Borough']=="Not assigned"].index, inplace = True) 
df.head()

Postal Code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

#### Dealing with multiple neighborhoods having same Postal Code

In [4]:
df1=df.groupby("Postal Code").agg(lambda x:','.join(set(x)))
df1=df1.reset_index()
df1.head()

Postal Code      Borough                            Neighborhood
0         M1B  Scarborough                          Malvern, Rouge
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

#### Dealing with Not assigned neighborhoods

In [5]:
#If a neighborhood name is empty and borough is non-empty, use the borough name as neighborhood name
df1.loc[df1['Neighborhood'].isnull(),'Neighborhood']=df1.loc[df1['Neighborhood'].isnull(),'Borough']
df1.head()

Postal Code      Borough                            Neighborhood
0         M1B  Scarborough                          Malvern, Rouge
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

#### Reading csv file containing the latitude and longitude data

In [6]:
df2=pd.read_csv('http://cocl.us/Geospatial_data')
df2.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

#### Combining both dataframes together (both are sorted by the Postal Codes)

In [7]:
df3= pd.merge(df1, df2)
df3.head()

Postal Code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

####  Working with only boroughs that contain the word Toronto

In [8]:
toronto_data= df3[df3['Borough'].str.contains("Toronto" or "toronto")].reset_index(drop=True)
toronto_data.head()

Postal Code          Borough                    Neighborhood   Latitude  \
0         M4E     East Toronto                     The Beaches  43.676357   
1         M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2         M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
3         M4M     East Toronto                 Studio District  43.659526   
4         M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

#### Marking these neighborhoods in the map of Toronto

In [9]:
map_toronto = folium.Map(location=[43.6532,-79.3832], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [10]:
#Foursquare credentials
# Cell code has been removed after storing the foursquare credentials to variables
CLIENT_ID='SBBS5CBGGQYPU2CNM2WOI44RIJP13JTNG13K3U40TFY1WZTD'
CLIENT_SECRET='IW3OHBMNSMK023B5OPBXHT5JXXORITXKEEKB3MQZPE2KQIRW'
VERSION = '20180605'

#### Function for getting nearby venues

In [11]:
def getNearbyVenues(names, latitudes, longitudes, radius=1000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [12]:
#Getting venues near all the neighbirhoods in Toronto
LIMIT=100
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

The Beaches
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West, Lawrence Park
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North & West, Forest Hill Road Park
The Annex, North Midtown, Yorkville
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Stn A PO Boxes
First Canadian Place, Underground city
Christie
Dufferin, Dovercourt Village
Little Portugal, Trinity
Brockton, Parkdale Village, Exhibition Place
High 

#### Checking the venues of Toronto we acquired

In [13]:
print(toronto_venues.shape)
toronto_venues.head()

(3189, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  The Beaches              43.676357              -79.293031   
1  The Beaches              43.676357              -79.293031   
2  The Beaches              43.676357              -79.293031   
3  The Beaches              43.676357              -79.293031   
4  The Beaches              43.676357              -79.293031   

               Venue  Venue Latitude  Venue Longitude  \
0  Glen Manor Ravine       43.676821       -79.293942   
1    Tori's Bakeshop       43.672114       -79.290331   
2  Beaches Bake Shop       43.680363       -79.289692   
3     The Beech Tree       43.680493       -79.288846   
4    The Fox Theatre       43.672801       -79.287272   

                  Venue Category  
0                          Trail  
1  Vegetarian / Vegan Restaurant  
2                         Bakery  
3                      Gastropub  
4            Indie Movie Theater

##### We see that we've acquired 3189 venues near the locations of our dataframe

#### One hot encoding of the venue categories

In [14]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Zoo  Accessories Store  Airport  Airport Lounge  American Restaurant  \
0    0                  0        0               0                    0   
1    0                  0        0               0                    0   
2    0                  0        0               0                    0   
3    0                  0        0               0                    0   
4    0                  0        0               0                    0   

   Amphitheater  Animal Shelter  Antique Shop  Aquarium  Art Gallery  ...  \
0             0               0             0         0            0  ...   
1             0               0             0         0            0  ...   
2             0               0             0         0            0  ...   
3             0               0             0         0            0  ...   
4             0               0             0         0            0  ...   

   University  Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0           0                              0                 0            0   
1           0                              1                 0            0   
2           0                              0                 0            0   
3           0                              0                 0            0   
4           0                              0                 0            0   

   Vietnamese Restaurant  Whisky Bar  Wine Bar  Wings Joint  Women's Store  \
0                      0           0         0            0              0   
1                      0           0         0            0              0   
2                      0           0         0            0              0   
3                      0           0         0            0              0   
4                      0           0         0            0              0   

   Yoga Studio  
0            0  
1            0  
2            0  
3            0  
4            0  

[5 rows x 279 columns]

In [15]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood   Zoo  \
0                                         Berczy Park  0.00   
1        Brockton, Parkdale Village, Exhibition Place  0.00   
2   Business reply mail Processing Centre, South C...  0.00   
3   CN Tower, King and Spadina, Railway Lands, Har...  0.00   
4                                  Central Bay Street  0.00   
5                                            Christie  0.00   
6                                Church and Wellesley  0.00   
7                      Commerce Court, Victoria Hotel  0.00   
8                                          Davisville  0.00   
9                                    Davisville North  0.00   
10                       Dufferin, Dovercourt Village  0.00   
11             First Canadian Place, Underground city  0.00   
12    Forest Hill North & West, Forest Hill Road Park  0.00   
13                           Garden District, Ryerson  0.00   
14  Harbourfront East, Union Station, Toronto Islands  0.00   
15                      High Park, The Junction South  0.00   
16                     India Bazaar, The Beaches West  0.00   
17          Kensington Market, Chinatown, Grange Park  0.00   
18                                      Lawrence Park  0.00   
19                           Little Portugal, Trinity  0.00   
20                        Moore Park, Summerhill East  0.00   
21                  North Toronto West, Lawrence Park  0.00   
22                             Parkdale, Roncesvalles  0.01   
23        Queen's Park, Ontario Provincial Government  0.00   
24                          Regent Park, Harbourfront  0.00   
25                           Richmond, Adelaide, King  0.00   
26                                           Rosedale  0.00   
27                                           Roselawn  0.00   
28                                 Runnymede, Swansea  0.00   
29                                     St. James Town  0.00   
30                        St. James Town, Cabbagetown  0.00   
31                                     Stn A PO Boxes  0.00   
32                                    Studio District  0.00   
33  Summerhill West, Rathnelly, South Hill, Forest...  0.00   
34                The Annex, North Midtown, Yorkville  0.00   
35                                        The Beaches  0.00   
36                       The Danforth West, Riverdale  0.00   
37           Toronto Dominion Centre, Design Exchange  0.00   
38                     University of Toronto, Harbord  0.00   

    Accessories Store   Airport  Airport Lounge  American Restaurant  \
0                0.00  0.000000        0.000000             0.010000   
1                0.01  0.000000        0.000000             0.010000   
2                0.00  0.000000        0.000000             0.020408   
3                0.00  0.066667        0.066667             0.000000   
4                0.00  0.000000        0.000000             0.010000   
5                0.00  0.000000        0.000000             0.010000   
6                0.00  0.000000        0.000000             0.010000   
7                0.00  0.000000        0.000000             0.020000   
8                0.00  0.000000        0.000000             0.000000   
9                0.00  0.000000        0.000000             0.000000   
10               0.00  0.000000        0.000000             0.000000   
11               0.00  0.000000        0.000000             0.020000   
12               0.00  0.000000        0.000000             0.000000   
13               0.00  0.000000        0.000000             0.010000   
14               0.00  0.000000        0.000000             0.000000   
15               0.00  0.000000        0.000000             0.000000   
16               0.00  0.000000        0.000000             0.000000   
17               0.00  0.000000        0.000000             0.000000   
18               0.00  0.000000        0.000000             0.000000   
19               0.00  0.000000        0.000000             0.010000   
20      

##### Now we have 274 unique venue categories, around 39 different neighborhoods of Toronto, and their relative presence in each neighborhood

#### For finding a neighborhood for our Indian Restaurant, we will mainly focus on 2 things for the primary analysis

* Neighborhoods with a large number of restaurants clustered around are more likely to be spots having a customer concentration, thus already have a potential market to tap into.
* If we can find the neighborhood with the highest number of restaurants, but WITHOUT an Indian restaurant already functioning nearby, we are finding a place where we don't have competition from the exact same cuisine

#### We will make another dataframe, which contains the data of only Restaurants as category

In [16]:
sample=toronto_grouped.filter(regex='Restaurant')

In [17]:
sample.head()

American Restaurant  Asian Restaurant  Belgian Restaurant  \
0             0.010000               0.0                 0.0   
1             0.010000               0.0                 0.0   
2             0.020408               0.0                 0.0   
3             0.000000               0.0                 0.0   
4             0.010000               0.0                 0.0   

   Brazilian Restaurant  Cajun / Creole Restaurant  Cantonese Restaurant  \
0                   0.0                        0.0                   0.0   
1                   0.0                        0.0                   0.0   
2                   0.0                        0.0                   0.0   
3                   0.0                        0.0                   0.0   
4                   0.0                        0.0                   0.0   

   Caribbean Restaurant  Chinese Restaurant  Comfort Food Restaurant  \
0                  0.00                0.00                     0.01   
1                  0.01                0.00                     0.01   
2                  0.00                0.00                     0.00   
3                  0.00                0.00                     0.00   
4                  0.00                0.01                     0.00   

   Cuban Restaurant  ...  Sushi Restaurant  Syrian Restaurant  \
0               0.0  ...          0.000000                0.0   
1               0.0  ...          0.000000                0.0   
2               0.0  ...          0.040816                0.0   
3               0.0  ...          0.000000                0.0   
4               0.0  ...          0.030000                0.0   

   Taiwanese Restaurant  Tapas Restaurant  Thai Restaurant  Theme Restaurant  \
0                   0.0              0.00         0.010000               0.0   
1                   0.0              0.01         0.000000               0.0   
2                   0.0              0.00         0.020408               0.0   
3                   0.0              0.00         0.000000               0.0   
4                   0.0              0.00         0.000000               0.0   

   Tibetan Restaurant  Turkish Restaurant  Vegetarian / Vegan Restaurant  \
0                0.00                 0.0                           0.01   
1                0.03                 0.0                           0.02   
2                0.00                 0.0                           0.00   
3                0.00                 0.0                           0.00   
4                0.00                 0.0                           0.02   

   Vietnamese Restaurant  
0                    0.0  
1                    0.0  
2                    0.0  
3                    0.0  
4                    0.0  

[5 rows x 53 columns]

In [18]:
sample.shape

(39, 53)

#### We've found 53 different kinds of Restaurants, let's see what they are

In [19]:
for item in sample.columns:
    print(item)

American Restaurant
Asian Restaurant
Belgian Restaurant
Brazilian Restaurant
Cajun / Creole Restaurant
Cantonese Restaurant
Caribbean Restaurant
Chinese Restaurant
Comfort Food Restaurant
Cuban Restaurant
Doner Restaurant
Dumpling Restaurant
Eastern European Restaurant
Ethiopian Restaurant
Falafel Restaurant
Fast Food Restaurant
Filipino Restaurant
French Restaurant
German Restaurant
Greek Restaurant
Hawaiian Restaurant
Indian Chinese Restaurant
Indian Restaurant
Indonesian Restaurant
Italian Restaurant
Japanese Restaurant
Jewish Restaurant
Korean Restaurant
Latin American Restaurant
Malay Restaurant
Mediterranean Restaurant
Mexican Restaurant
Middle Eastern Restaurant
Modern European Restaurant
New American Restaurant
North Indian Restaurant
Pakistani Restaurant
Persian Restaurant
Portuguese Restaurant
Ramen Restaurant
Restaurant
Seafood Restaurant
South American Restaurant
Sushi Restaurant
Syrian Restaurant
Taiwanese Restaurant
Tapas Restaurant
Thai Restaurant
Theme Restaurant
Tibeta

#### Now, we remove the Indian restaurant column from this dataframe

In [20]:
sample.drop(['Indian Restaurant'], axis = 1,inplace=True)

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/pandas/core/frame.py:3997: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [21]:
sample.head()

American Restaurant  Asian Restaurant  Belgian Restaurant  \
0             0.010000               0.0                 0.0   
1             0.010000               0.0                 0.0   
2             0.020408               0.0                 0.0   
3             0.000000               0.0                 0.0   
4             0.010000               0.0                 0.0   

   Brazilian Restaurant  Cajun / Creole Restaurant  Cantonese Restaurant  \
0                   0.0                        0.0                   0.0   
1                   0.0                        0.0                   0.0   
2                   0.0                        0.0                   0.0   
3                   0.0                        0.0                   0.0   
4                   0.0                        0.0                   0.0   

   Caribbean Restaurant  Chinese Restaurant  Comfort Food Restaurant  \
0                  0.00                0.00                     0.01   
1                  0.01                0.00                     0.01   
2                  0.00                0.00                     0.00   
3                  0.00                0.00                     0.00   
4                  0.00                0.01                     0.00   

   Cuban Restaurant  ...  Sushi Restaurant  Syrian Restaurant  \
0               0.0  ...          0.000000                0.0   
1               0.0  ...          0.000000                0.0   
2               0.0  ...          0.040816                0.0   
3               0.0  ...          0.000000                0.0   
4               0.0  ...          0.030000                0.0   

   Taiwanese Restaurant  Tapas Restaurant  Thai Restaurant  Theme Restaurant  \
0                   0.0              0.00         0.010000               0.0   
1                   0.0              0.01         0.000000               0.0   
2                   0.0              0.00         0.020408               0.0   
3                   0.0              0.00         0.000000               0.0   
4                   0.0              0.00         0.000000               0.0   

   Tibetan Restaurant  Turkish Restaurant  Vegetarian / Vegan Restaurant  \
0                0.00                 0.0                           0.01   
1                0.03                 0.0                           0.02   
2                0.00                 0.0                           0.00   
3                0.00                 0.0                           0.00   
4                0.00                 0.0                           0.02   

   Vietnamese Restaurant  
0                    0.0  
1                    0.0  
2                    0.0  
3                    0.0  
4                    0.0  

[5 rows x 52 columns]

##### We find that the number of column has been reduced by 1, the Indian restaurant column which we removed

#### Now we rejoin the neighborhood column to this dataframe

In [22]:
sample['Neighborhood'] =toronto_grouped['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [sample.columns[-1]] + list(sample.columns[:-1])
sample = sample[fixed_columns]

sample.head()

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


Neighborhood  American Restaurant  \
0                                        Berczy Park             0.010000   
1       Brockton, Parkdale Village, Exhibition Place             0.010000   
2  Business reply mail Processing Centre, South C...             0.020408   
3  CN Tower, King and Spadina, Railway Lands, Har...             0.000000   
4                                 Central Bay Street             0.010000   

   Asian Restaurant  Belgian Restaurant  Brazilian Restaurant  \
0               0.0                 0.0                   0.0   
1               0.0                 0.0                   0.0   
2               0.0                 0.0                   0.0   
3               0.0                 0.0                   0.0   
4               0.0                 0.0                   0.0   

   Cajun / Creole Restaurant  Cantonese Restaurant  Caribbean Restaurant  \
0                        0.0                   0.0                  0.00   
1                        0.0                   0.0                  0.01   
2                        0.0                   0.0                  0.00   
3                        0.0                   0.0                  0.00   
4                        0.0                   0.0                  0.00   

   Chinese Restaurant  Comfort Food Restaurant  ...  Sushi Restaurant  \
0                0.00                     0.01  ...          0.000000   
1                0.00                     0.01  ...          0.000000   
2                0.00                     0.00  ...          0.040816   
3                0.00                     0.00  ...          0.000000   
4                0.01                     0.00  ...          0.030000   

   Syrian Restaurant  Taiwanese Restaurant  Tapas Restaurant  Thai Restaurant  \
0                0.0                   0.0              0.00         0.010000   
1                0.0                   0.0              0.01         0.000000   
2                0.0                   0.0              0.00         0.020408   
3                0.0                   0.0              0.00         0.000000   
4                0.0                   0.0              0.00         0.000000   

   Theme Restaurant  Tibetan Restaurant  Turkish Restaurant  \
0               0.0                0.00                 0.0   
1               0.0                0.03                 0.0   
2               0.0                0.00                 0.0   
3               0.0                0.00                 0.0   
4               0.0                0.00                 0.0   

   Vegetarian / Vegan Restaurant  Vietnamese Restaurant  
0                           0.01                    0.0  
1                           0.02                    0.0  
2                           0.00                    0.0  
3                           0.00                    0.0  
4                           0.02                    0.0  

[5 rows x 53 columns]

#### Now we sum the rows horizontally, to get an aggregate value for each Neighborhood, representing the aggregation of all Non Indian Restaurants

In [23]:
aggregate=sample.sum(axis=1)

In [24]:
aggregate.head()

0    0.180000
1    0.250000
2    0.204082
3    0.000000
4    0.240000
dtype: float64

In [25]:
sample2= pd.DataFrame(columns=['Neighborhood','Aggregate Restaurants','Indian Restaurants'])

In [26]:
sample2['Neighborhood']=sample['Neighborhood']
sample2['Aggregate Restaurants']=aggregate

In [27]:
sample2.head()

Neighborhood  Aggregate Restaurants  \
0                                        Berczy Park               0.180000   
1       Brockton, Parkdale Village, Exhibition Place               0.250000   
2  Business reply mail Processing Centre, South C...               0.204082   
3  CN Tower, King and Spadina, Railway Lands, Har...               0.000000   
4                                 Central Bay Street               0.240000   

  Indian Restaurants  
0                NaN  
1                NaN  
2                NaN  
3                NaN  
4                NaN

In [28]:
sample2['Indian Restaurants']=toronto_grouped['Indian Restaurant']

In [29]:
sample2.head()

Neighborhood  Aggregate Restaurants  \
0                                        Berczy Park               0.180000   
1       Brockton, Parkdale Village, Exhibition Place               0.250000   
2  Business reply mail Processing Centre, South C...               0.204082   
3  CN Tower, King and Spadina, Railway Lands, Har...               0.000000   
4                                 Central Bay Street               0.240000   

   Indian Restaurants  
0                0.00  
1                0.02  
2                0.00  
3                0.00  
4                0.00

In [30]:
sample2

Neighborhood  Aggregate Restaurants  \
0                                         Berczy Park               0.180000   
1        Brockton, Parkdale Village, Exhibition Place               0.250000   
2   Business reply mail Processing Centre, South C...               0.204082   
3   CN Tower, King and Spadina, Railway Lands, Har...               0.000000   
4                                  Central Bay Street               0.240000   
5                                            Christie               0.310000   
6                                Church and Wellesley               0.230000   
7                      Commerce Court, Victoria Hotel               0.230000   
8                                          Davisville               0.340000   
9                                    Davisville North               0.250000   
10                       Dufferin, Dovercourt Village               0.233333   
11             First Canadian Place, Underground city               0.220000   
12    Forest Hill North & West, Forest Hill Road Park               0.250000   
13                           Garden District, Ryerson               0.270000   
14  Harbourfront East, Union Station, Toronto Islands               0.150000   
15                      High Park, The Junction South               0.220000   
16                     India Bazaar, The Beaches West               0.177215   
17          Kensington Market, Chinatown, Grange Park               0.230000   
18                                      Lawrence Park               0.000000   
19                           Little Portugal, Trinity               0.320000   
20                        Moore Park, Summerhill East               0.296875   
21                  North Toronto West, Lawrence Park               0.239130   
22                             Parkdale, Roncesvalles               0.210000   
23        Queen's Park, Ontario Provincial Government               0.260000   
24                          Regent Park, Harbourfront               0.160000   
25                           Richmond, Adelaide, King               0.190000   
26                                           Rosedale               0.095238   
27                                           Roselawn               0.227273   
28                                 Runnymede, Swansea               0.192308   
29                                     St. James Town               0.230000   
30                        St. James Town, Cabbagetown               0.297297   
31                                     Stn A PO Boxes               0.210000   
32                                    Studio District               0.270000   
33  Summerhill West, Rathnelly, South Hill, Forest...               0.296296   
34                The Annex, North Midtown, Yorkville               0.270000   
35                                        The Beaches               0.181818   
36                       The Danforth West, Riverdale               0.320000   
37           Toronto Dominion Centre, Design Exchange               0.220000   
38                     University of Toronto, Harbord               0.250000   

    Indian Restaurants  
0             0.000000  
1             0.020000  
2             0.000000  
3             0.000000  
4             0.000000  
5             0.020000  
6             0.010000  
7             0.000000  
8             0.030000  
9             0.000000  
10            0.000000  
11            0.000000  
12            0.000000  
13            0.000000  
14            0.010000  
15            0.010000  
16            0.088608  
17            0.000000  
18            0.000000  
19            0.000000  
20            0.000000  
21            0.000000  
22            0.000000  
23            0.010000  
24            0.020000  
25            0.000000  
26            0.000000  
27            0.000000  
28            0.012821  
29            0.000000  
30            0.027027  
31            0.000000  
32            0

#### This is the complete list of neighborhoods in Toronto with an aggregated representation of all non Indian restaurants, and just Indian restaurants as separate columns

#### Rows with higher number in the aggregate column shows the neighborhoods with a clustering of restaurants, and rows having 0 value for Indian restaurants shows neighborhoods with no Indian restaurant

In [31]:
#Sorting the database in decreasing order of the aggregate values
sample2.sort_values(by=['Aggregate Restaurants'], inplace=True, ascending=False)


In [32]:
sample2.head()

Neighborhood  Aggregate Restaurants  Indian Restaurants
8                     Davisville               0.340000            0.030000
36  The Danforth West, Riverdale               0.320000            0.010000
19      Little Portugal, Trinity               0.320000            0.000000
5                       Christie               0.310000            0.020000
30   St. James Town, Cabbagetown               0.297297            0.027027

In [33]:
sample2.head(10)

Neighborhood  Aggregate Restaurants  \
8                                          Davisville               0.340000   
36                       The Danforth West, Riverdale               0.320000   
19                           Little Portugal, Trinity               0.320000   
5                                            Christie               0.310000   
30                        St. James Town, Cabbagetown               0.297297   
20                        Moore Park, Summerhill East               0.296875   
33  Summerhill West, Rathnelly, South Hill, Forest...               0.296296   
13                           Garden District, Ryerson               0.270000   
32                                    Studio District               0.270000   
34                The Annex, North Midtown, Yorkville               0.270000   

    Indian Restaurants  
8             0.030000  
36            0.010000  
19            0.000000  
5             0.020000  
30            0.027027  
20            0.000000  
33            0.000000  
13            0.000000  
32            0.000000  
34            0.020000

#### From the above data frame, let's consider the top 5 rows, we see that that Davisville has the highest clustering of restaurants around, and it also has the highest number of Indian restaurants around. In fact, all neighborhoods in the top 5, except Little Portugal has an Indian restaurant operating nearby.

#### We can also see that Little Portugal and the Danforth West has the same value for aggregate restaurants, making it a tie for second place, but makes it a better option to start our Indian restaurant. Hence, we can focus more on the Little Portugal neighborhood

#### One important task left before us now is making sure that the neighborhood we have selected is cosmopolitan in nature. We don't want to take a risk of residents of the area not being interested to try out a new cuisine 

#### We will now see what other cuisines are available around the Little Portugal region 

In [34]:
sample3=sample.loc[sample['Neighborhood'] == "Little Portugal, Trinity"]

In [35]:
sample3

Neighborhood  American Restaurant  Asian Restaurant  \
19  Little Portugal, Trinity                 0.01              0.04   

    Belgian Restaurant  Brazilian Restaurant  Cajun / Creole Restaurant  \
19                 0.0                   0.0                        0.0   

    Cantonese Restaurant  Caribbean Restaurant  Chinese Restaurant  \
19                   0.0                   0.0                 0.0   

    Comfort Food Restaurant  ...  Sushi Restaurant  Syrian Restaurant  \
19                      0.0  ...               0.0                0.0   

    Taiwanese Restaurant  Tapas Restaurant  Thai Restaurant  Theme Restaurant  \
19                   0.0              0.01             0.01               0.0   

    Tibetan Restaurant  Turkish Restaurant  Vegetarian / Vegan Restaurant  \
19                 0.0                 0.0                           0.05   

    Vietnamese Restaurant  
19                   0.01  

[1 rows x 53 columns]

#### Now we try to find the various cuisines of restaurants operating in the Little Portugal Neighborhood, although from the first look, we see that it has American, Asian,Tapas,Thai,Vegan and Vietnamese cuisines, which would already make it a multi cuisine neighborhood, and that should play in favor of the stakeholder

In [36]:
i=0
for column in sample3.columns:
    if sample3.iloc[0,i]!=0:
        print(column)
    i+=1
        

Neighborhood
American Restaurant
Asian Restaurant
Cuban Restaurant
Dumpling Restaurant
French Restaurant
Greek Restaurant
Italian Restaurant
Japanese Restaurant
Korean Restaurant
Malay Restaurant
New American Restaurant
Restaurant
Seafood Restaurant
Tapas Restaurant
Thai Restaurant
Vegetarian / Vegan Restaurant
Vietnamese Restaurant


#### We see that the neighborhood has all the above cuisines and a great spot to find a wide variety of cuisines from all over the world, this would rule out the stakeholder's doubt whether a new cuisine will be welcome there, and we can select Little Portugal to be a neighborhood of our choice based on this primary analysis

#### Now, let us plot the Little Portugal neighborhood on the map of Toronto

In [41]:
lat=float(toronto_data.loc[toronto_data['Neighborhood']=='Little Portugal, Trinity','Latitude'])
lng=float(toronto_data.loc[toronto_data['Neighborhood']=='Little Portugal, Trinity','Longitude'])

In [45]:
map_new = folium.Map(location=[43.6532,-79.3832], zoom_start=12)

folium.Marker(
    location=[lat,lng],
    popup='LIttle Portugal'
).add_to(map_new)
map_new

From the primary analysis, we have chosen <b>Little Portugal</b> as the preferred neighborhood for starting the new Indian Restaurant. The final decision has to be made only after considering other factors like availability of an ideal space for rent, local traffic, and extra amenities